Libraries/LibCrypto/PK/RSA.h

```diff
@@ -0,0 +1,234 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Vector.h>
+#include <LibCrypto/BigInt/UnsignedBigInteger.h>
+#include <LibCrypto/NumberTheory/ModularFunctions.h>
+#include <LibCrypto/PK/Code/EMSA_PSS.h>
+#include <LibCrypto/PK/PK.h>
+
+namespace Crypto {
+namespace PK {
+template<typename Integer = u64>
+class RSAPublicKey {
+public:
+    RSAPublicKey(const Integer& n, const Integer& e)
+        : m_modulus(n)
+        , m_public_exponent(e)
+    {
+    }
+
+    RSAPublicKey()
+        : m_modulus(0)
+        , m_public_exponent(0)
+    {
+    }
+
+    //--stuff it should do
+
+    const Integer& modulus() const { return m_modulus; }
+    const Integer& public_exponent() const { return m_public_exponent; }
+    size_t length() const { return m_length; }
+    void set_length(size_t length) { m_length = length; }
+
+    void set(const Integer& n, const Integer& e)
+    {
+        m_modulus = n;
+        m_public_exponent = e;
+        m_length = (n.trimmed_length() * sizeof(u32));
+    }
+
+private:
+    Integer m_modulus;
+    Integer m_public_exponent;
+    size_t m_length { 0 };
+};
+
+template<typename Integer = UnsignedBigInteger>
+class RSAPrivateKey {
+public:
+    RSAPrivateKey(const Integer& n, const Integer& d, const Integer& e)
+        : m_modulus(n)
+        , m_private_exponent(d)
+        , m_public_exponent(e)
+    {
+    }
+
+    RSAPrivateKey()
+    {
+    }
+
+    //--stuff it should do
+    const Integer& modulus() const { return m_modulus; }
+    const Integer& private_exponent() const { return m_private_exponent; }
+    const Integer& public_exponent() const { return m_public_exponent; }
+    size_t length() const { return m_length; }
+    void set_length(size_t length) { m_length = length; }
+
+    void set(const Integer& n, const Integer& d, const Integer& e)
+    {
+        m_modulus = n;
+        m_private_exponent = d;
+        m_public_exponent = e;
+        m_length = (n.length() * sizeof(u32));
+    }
+
+private:
+    Integer m_modulus;
+    Integer m_private_exponent;
+    Integer m_public_exponent;
+    size_t m_length { 0 };
+};
+
+template<typename PubKey, typename PrivKey>
+struct RSAKeyPair {
+    PubKey public_key;
+    PrivKey private_key;
+};
```

Comment:
if I'm not mistaken, the RSA private key contains all the data of the public key. Does it make sense to store both in independent objects ?

Reply:
Yes, we might have a public key but no private key.

The data stored here is actually a subset of all the data, since we don't make much use of the private key, I opted to have it barebone.